dev-tools/ls-apis/src/workspaces.rs

```diff
@@ -0,0 +1,270 @@
+        // one in the Omicron repo.
+        if server_pkgname == "dpd-client" && found_in_workspaces.len() == 2 {
+            if found_in_workspaces[0].0.name() == "omicron" {
+                return Ok(found_in_workspaces[0]);
+            }
+            if found_in_workspaces[1].0.name() == "omicron" {
+                return Ok(found_in_workspaces[1]);
+            }
+        }
+        ensure!(
+            !found_in_workspaces.is_empty(),
+            "server package {:?} was not found in any workspace",
+            server_pkgname
+        );
+        ensure!(
+            found_in_workspaces.len() == 1,
+            "server package {:?} was found in more than one workspace: {}",
+            server_pkgname,
+            found_in_workspaces
+                .into_iter()
+                .map(|(w, _)| w.name())
+                .collect::<Vec<_>>()
+                .join(", ")
+        );
+        Ok(found_in_workspaces[0])
+    }
+
+    /// Returns the set of distinct pkgids for package "pkg" among all
+    /// workspaces.
+    pub fn workspace_pkgids<'a>(
+        &'a self,
+        pkgname: &'a str,
+    ) -> BTreeSet<&'a PackageId> {
+        self.workspaces.values().flat_map(move |w| w.pkgids(pkgname)).collect()
+    }
+}
+
+/// Load a `Workspace` for a repo `repo` using the manifest path inferred by
+/// looking up one of its packages `pkgname` in `workspace`
+///
+/// For example, we might locate the Crucible repo by looking up the
+/// `crucible-pantry-client` package in the Omicron workspace, finding its
+/// manifest path, and locating the containing Crucible workspace.
+fn load_dependent_repo(
+    workspace: &Workspace,
+    repo: &str,
+    pkgname: &str,
+) -> Result<Workspace> {
+    // `Workspace` doesn't let us look up a non-workspace package by name
+    // because there may be many of them.  So list all the pkgids and take any
+    // one of them -- any of them should work for our purpoes.
+    let pkgid = workspace.pkgids(pkgname).next().ok_or_else(|| {
+        anyhow!(
+            "workspace {} did not contain expected package {}",
+            workspace.name(),
+            pkgname
+        )
+    })?;
+
+    // Now we can look up the package metadata.
+    let pkg = workspace.pkg_by_id(pkgid).ok_or_else(|| {
+        anyhow!(
+            "workspace {}: did not contain expected package id {}",
+            workspace.name(),
+            pkgname
+        )
+    })?;
+
+    // The package metadata should show where the package's manifest file should
+    // be.  This may be buried deep in the workspace.  How do we find the root
+    // of the workspace?  We assume (and verify) that the workspace is checked
+    // out in the usual place under CARGO_HOME.  Then the workspace root is the
+    // full path two component deeper than the checkouts directory.
+    let cargo_home = std::env::var("CARGO_HOME").context("CARGO_HOME")?;
+    let git_checkouts: Utf8PathBuf =
```

Comment:
A better way to do this would be `cargo locate-project --workspace --manifest-path {pkg.manifest_path} --message-format plain`.

Reply:
Thanks!  Done in 358a5e477.